state/cluster/badger/mutator.go

```diff
@@ -36,202 +37,286 @@ func NewMutableState(state *State, tracer module.Tracer, headers storage.Headers
 	return mutableState, nil
 }
 
+// extendContext encapsulates all state information required in order to validate a candidate cluster block.
+type extendContext struct {
+	candidate                *cluster.Block // the proposed candidate cluster block
+	finalizedClusterBlock    flow.Header    // the latest finalized cluster block
+	finalizedConsensusHeight uint64         // the latest finalized height on the main change
+	epochFirstHeight         uint64         // the first height of this cluster's operating epoch
+	epochLastHeight          uint64         // the last height of this cluster's operating epoch (may be unknown)
+	epochHasEnded            bool           // whether this cluster's operating epoch has ended (whether the above field is known)
+}
+
+// getExtendCtx reads all required information from the database in order to validate
+// a candidate extending cluster block.
+// No errors are expected during normal operation.
+func (m *MutableState) getExtendCtx(candidate *cluster.Block) (extendContext, error) {
+	var ctx extendContext
+	ctx.candidate = candidate
+
+	err := m.State.db.View(func(tx *badger.Txn) error {
+		// get the latest finalized cluster block and latest finalized consensus height
+		err := procedure.RetrieveLatestFinalizedClusterHeader(candidate.Header.ChainID, &ctx.finalizedClusterBlock)(tx)
+		if err != nil {
+			return fmt.Errorf("could not retrieve finalized cluster head: %w", err)
+		}
+		err = operation.RetrieveFinalizedHeight(&ctx.finalizedConsensusHeight)(tx)
+		if err != nil {
+			return fmt.Errorf("could not retrieve finalized height on consensus chain: %w", err)
+		}
+
+		err = operation.RetrieveEpochFirstHeight(m.State.epoch, &ctx.epochFirstHeight)(tx)
+		if err != nil {
+			return fmt.Errorf("could not get operating epoch first height: %w", err)
+		}
+		err = operation.RetrieveEpochLastHeight(m.State.epoch, &ctx.epochLastHeight)(tx)
+		if err != nil {
+			if errors.Is(err, storage.ErrNotFound) {
+				ctx.epochHasEnded = false
+				return nil
+			}
+			return fmt.Errorf("unexpected failure to retrieve final height of operating epoch: %w", err)
+		}
+		ctx.epochHasEnded = true
+		return nil
+	})
+	if err != nil {
+		return extendContext{}, fmt.Errorf("could not read required state information for Extend checks: %w", err)
+	}
+	return ctx, nil
+}
+
 // Extend validates that the given cluster block passes compliance rules, then inserts
 // it to the cluster state.
 // TODO (Ramtin) pass context here
 // Expected errors during normal operations:
 //   - state.OutdatedExtensionError if the candidate block is outdated (e.g. orphaned)
+//   - state.UnverifiableExtensionError if the candidate block cannot be verified
```

Comment:
would it be correct to say:
```suggestion
//   - state.UnverifiableExtensionError if the reference block is _not_ a known finalized block
```
Trying to be as concrete as possible. If you agree, could you also update the interface doc 🙏 🙇